pandas/_libs/lib.pyx

```diff
@@ -2449,6 +2450,18 @@ def maybe_convert_numeric(
         elif is_decimal(val):
             floats[i] = complexes[i] = val
             seen.float_ = True
+        elif convert_datetime and (PyDate_Check(val) or cnp.is_datetime64_object(val)):
+            # convert_datetime flag avoids conversion for base_readers
+            # PyDate_Check also includes PyDatetime_Check
+            seen.datetime_ = True
+            if val in na_values or checknull(val):
+                seen.saw_null()
+                mask[i] = 1
+                floats[i] = NaN
+            else:
+                ints[i] = np.datetime64(val).astype(np.int64)
+                # because of pd.NaT, we may need to return in floats #GH 42380
+                floats[i] = <float64_t>ints[i]
```

Comment:
We definitely don't want to cast to float here.

This issue is really challenging and probably impossible to solve using the default (NumPy-backed) type system of pandas. I'm not sure there's anything to do unless that changes

Reply:
@WillAyd, Casting to float in addition to other dtypes is the existing pattern in `maybe_convert_numeric` though.. The pattern is for null values, the float array will be returned if there is null values in the series. 

e.g., 
https://github.com/pandas-dev/pandas/blob/82a0a2dab77cb503dcfbefd3a970aa0a6820e0cd/pandas/_libs/lib.pyx#L2411

https://github.com/pandas-dev/pandas/blob/82a0a2dab77cb503dcfbefd3a970aa0a6820e0cd/pandas/_libs/lib.pyx#L2429